lib/cadet/assessments.ex

```diff
@@ -0,0 +1,590 @@
+        previous =
+          Enum.find(
+            submission.answers,
+            &(&1.id == String.to_integer(ps["id"]))
+          )
+
+        changeset = Answer.changeset(previous, ps)
+        Repo.update!(changeset)
+      end
+
+      total_marks =
+        params["answers"]
+        |> Enum.map(fn {_, params} -> params["marks"] end)
+        |> Enum.map(fn mark ->
+          {parsed, _} = Float.parse(mark)
+          parsed
+        end)
+        |> Enum.sum()
+
+      max_marks =
+        submission.answers
+        |> Enum.map(& &1.question.weight)
+        |> Enum.sum()
+
+      total_marks = Enum.min([total_marks, max_marks])
+      max_xp = submission.mission.max_xp
+      previous_xp = previous_marks / max_marks * max_xp
+      total_xp = total_marks / max_marks * max_xp
+      to_subtract = submission.override_xp || previous_xp
+
+      to_add =
+        if params["override_xp"] == "" do
+          total_xp
+        else
+          String.to_integer(params["override_xp"])
+        end
+
+      delta_xp = round(to_add - to_subtract)
+      update = if previous_marks > 0, do: " update", else: ""
+
+      # Increase XP
+      mission = submission.mission
+      name = mission.name
+
+      {:ok, xp_history} =
+        Course.create_xp_history(
+          %{
+            "reason" => "#{name} XP#{update}. (#{total_marks}/#{max_marks})",
+            "amount" => delta_xp
+          },
+          submission.student_id,
+          grader.id
+        )
+
+      # Set submission to graded
+      submission
+      |> Submission.changeset(params)
+      |> change(%{status: :graded})
+      |> Repo.update!()
+
+      submission
+    end)
+  end
+
+  def prepare_game(student) do
+    # Get all non-attempted mission
+    missions =
+      Repo.all(
+        from(
+          a in Mission,
+          left_join: s in Submission,
+          on: a.id == s.mission_id and s.student_id == ^student.id,
+          where: is_nil(s.student_id),
+          select: a
+        )
+      )
+
+    # Filter those can be attempted, min by order, mission first
+    missions =
+      missions
+      |> Enum.filter(&(&1.type != :path && can_attempt?(&1, student.user)))
+
+    if Enum.empty?(missions) do
+      nil
+    else
+      [hd | _] = Enum.sort_by(missions, &story_name_pair/1, &compare_story_name_pair/2)
+      hd
+    end
+  end
+
+  def story_name_pair(mission) do
+    {mission.type, mission.name}
+  end
+
+  def compare_story_name_pair({t1, n1}, {t2, n2}) do
+    cond do
+      t1 == t2 -> n1 <= n2
+      t1 == :mission -> true
+      t1 == :sidequest && t2 == :mission -> false
+      t1 == :contest && t2 == :mission -> false
+      t1 == :contest && t2 == :sidequest -> false
+    end
+  end
+
+  def story_name(mission) do
+    type = Atom.to_string(mission.type)
+    String.downcase(type) <> "-" <> mission.name
+  end
+
+  defp create_blank_question(question, type) do
+  end
+
+  defp create_concrete_question(question, type) do
```

Comment:
I don't think we need this since the question type polymorphism is in the JSON. @evansb what's your opinion?